File: app/src/main/java/com/corphish/nightlight/Helpers/AlarmUtils.java
package com.corphish.nightlight.Helpers;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;

import com.corphish.nightlight.Receivers.StartNLReceiver;
import com.corphish.nightlight.Receivers.StopNLReceiver;

import java.util.Calendar;

/**
 * Created by Avinaba on 10/6/2017.
 * Helper class to deal with alarms
 */

public class AlarmUtils {

    private static final int REQUEST_CODE_START     = 0;
    private static final int REQUEST_CODE_STOP      = 0;

    /**
     * Sets the start and end alarms on user specified time
     * @param context - Needed by intent, pendingIntent and to get the AlarmManager service
     * @param startTime - Starting time for alarm
     * @param endTime - Ending time for alarm
     */
    public static void setAlarms(Context context, String startTime, String endTime) {
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        long timeInMillis;

        Intent startIntent = new Intent(context, StartNLReceiver.class);
        PendingIntent startAlarmIntent = PendingIntent.getBroadcast(context, REQUEST_CODE_START, startIntent, 0);

        Intent endIntent = new Intent(context, StopNLReceiver.class);
        PendingIntent endAlarmIntent = PendingIntent.getBroadcast(context, REQUEST_CODE_STOP, endIntent, 0);

        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        calendar.set(Calendar.HOUR_OF_DAY, TimeUtils.getTimeAsHourAndMinutes(startTime)[0]);
        calendar.set(Calendar.MINUTE, TimeUtils.getTimeAsHourAndMinutes(startTime)[1]);

        timeInMillis = calendar.getTimeInMillis();
        if (TimeUtils.getCurrentTimeAsMinutes() > TimeUtils.getTimeInMinutes(startTime)) timeInMillis += 86400000L;

        alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, timeInMillis, AlarmManager.INTERVAL_DAY, startAlarmIntent);

        calendar.setTimeInMillis(System.currentTimeMillis());
        calendar.set(Calendar.HOUR_OF_DAY, TimeUtils.getTimeAsHourAndMinutes(endTime)[0]);
        calendar.set(Calendar.MINUTE, TimeUtils.getTimeAsHourAndMinutes(endTime)[1]);

        timeInMillis = calendar.getTimeInMillis();
        if (TimeUtils.getCurrentTimeAsMinutes() > TimeUtils.getTimeInMinutes(endTime)) timeInMillis += 86400000L;

        alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, timeInMillis, AlarmManager.INTERVAL_DAY, endAlarmIntent);
    }
}

File: app/src/main/java/com/corphish/nightlight/helpers/AlarmUtils.kt
package com.corphish.nightlight.helpers

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import com.corphish.nightlight.receivers.AutomateSignalReceiver

import java.util.Calendar

/**
 * Created by Avinaba on 10/6/2017.
 * Helper class to deal with alarms.
 *
 * As part of 3.0 update in 2020, alarms set will no longer will be
 * repeating, as only one alarm will be set at any given time. When this alarm
 * is fired, it will set subsequent alarms accordingly.
 */

object AlarmUtils {
    // Request code
    private const val REQUEST_CODE = 42

    /**
     * Sets a non-repeating alarm in a given time of day.
     * If the time specified is lesser than current time, alarm will be set for
     * next day.
     *
     * @param context Context.
     * @param time    Absolute time.
     */
    fun setAlarmAbsolute(context: Context, time: String) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, AutomateSignalReceiver::class.java)
        val alarmIntent = PendingIntent.getBroadcast(context, REQUEST_CODE, intent, 0)

        val calendar = Calendar.getInstance()
        val splitTime = TimeUtils.getTimeAsHourAndMinutes(time)
        calendar.timeInMillis = System.currentTimeMillis()
        calendar.set(Calendar.HOUR_OF_DAY, splitTime[0])
        calendar.set(Calendar.MINUTE, splitTime[1])

        var timeInMillis = calendar.timeInMillis
        if (TimeUtils.currentTimeAsMinutes > TimeUtils.getTimeInMinutes(time)) {
            timeInMillis += 86400000L
        }

        alarmManager.set(AlarmManager.RTC_WAKEUP, timeInMillis, alarmIntent)
    }

    /**
     * Sets a non-repeating alarm after given minutes from current time.
     *
     * @param context Context.
     * @param minutes Minutes after which alarm must be set.
     */
    fun setAlarmRelative(context: Context, minutes: Int) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, AutomateSignalReceiver::class.java)
        val alarmIntent = PendingIntent.getBroadcast(context, REQUEST_CODE, intent, 0)

        val targetMillis = System.currentTimeMillis() + (minutes * 60L * 1000)

        alarmManager.set(AlarmManager.RTC_WAKEUP, targetMillis, alarmIntent)
    }
}
